compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,274 @@
+            extensionReceiverParameter = function.extensionReceiverParameter?.copyWithTypeErasure(this, typeMap)
+            function.valueParameters.mapIndexed { i, param ->
+                valueParameters.add(i, param.copyWithTypeErasure(this, typeMap))
+            }
+            val exception = context.getTopLevelClass(FqName("java.lang.UnsupportedOperationException"))
+            // Function body consist only of throwing UnsupportedOperationException statement
+            body = context.createIrBuilder(function.symbol).irBlockBody {
+                +irThrow(
+                    irCall(
+                        exception.owner.constructors.single {
+                            it.valueParameters.size == 1 && it.valueParameters.single().type.isNullableString()
+                        }
+                    ).apply {
+                        putValueArgument(0, irString("Operation is not supported for read-only collection"))
+                    }
+                )
+            }
+        }
+    }
+
+    // Type erasure for value parameters, with type substitution
+    private fun IrValueParameter.copyWithTypeErasure(target: IrSimpleFunction, typeMap: Map<IrTypeParameterSymbol, IrType>)
+            : IrValueParameter {
+        val descriptor = WrappedValueParameterDescriptor(this.descriptor.annotations)
+        return IrValueParameterImpl(
+            UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+            IrDeclarationOrigin.IR_BUILTINS_STUB,
+            IrValueParameterSymbolImpl(descriptor),
+            name,
+            index,
+            type.substitute(typeMap).eraseTypeParameters(),
+            varargElementType?.substitute(typeMap)?.eraseTypeParameters(),
+            isCrossinline,
+            isNoinline
+        ).apply {
+            descriptor.bind(this)
+            parent = target
+        }
+    }
+
+    // Create a map for type parameters between source class (Mutable Collection classes) to
+    // target class (class currently in lowering phase), this map is later used for substituting type parameters in generated functions
+    private fun createTypeMap(collectionClass: IrClass, mutableClass: IrClass, targetClass: IrClass): Map<IrTypeParameterSymbol, IrType> {
+        // It is guaranteed by compiler that a class inherits from an immutable collection class must have
+        // some methods overridden, therefore we search for such overriddens, and get substitute type for
+        // our stubs based on these functions. For every collection class we select specific function(s), to make sure all
+        // type parameters are covered by our selection.
+        val targetTypeList = when (collectionClass.name.asString()) {
+            "Iterator", "ListIterator" -> {
+                listOf("next")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
```

Comment:
Please collapse successive `.map` calls into one for readability and to avoid extra allocations, e.g.:
```
    listOf("next")
        .map { name -> getFunctionForSubstitution(name, collectionClass, targetClass).returnType }
```